src/Parser/MarksmanshipHunter/Modules/Items/Tier19_2p.js

```diff
@@ -0,0 +1,70 @@
+import React from 'react';
+import Module from 'Parser/Core/Module';
+import Combatants from 'Parser/Core/Modules/Combatants';
+import SpellUsable from 'Parser/Core/Modules/SpellUsable';
+import RESOURCE_TYPES from 'common/RESOURCE_TYPES';
+
+import SPELLS from 'common/SPELLS';
+import SpellIcon from 'common/SpellIcon';
+import SpellLink from 'common/SpellLink';
+import { formatNumber } from "../../../../common/format";
+
+const debug = false;
+
+const CDR_PER_FOCUS = 22.22222;
+
+class Tier19_2p extends Module {
+  static dependencies = {
+    combatants: Combatants,
+    spellUsable: SpellUsable,
+  };
+
+  effectiveTrueshotReductionMs = 0;
+  wastedTrueshotReductionMs = 0;
+
+  focusUpdates = [];
+
+  on_initialized() {
+    this.active = this.combatants.selected.hasBuff(SPELLS.HUNTER_MM_T19_2P_BONUS.id);
+  }
+  on_byPlayer_cast(event) {
+    const spellId = event.ability.guid;
+    if (spellId !== SPELLS.AIMED_SHOT.id && spellId !== SPELLS.PIERCING_SHOT_TALENT.id && spellId !== SPELLS.BURSTING_SHOT.id && spellId !== SPELLS.MARKED_SHOT.id && spellId !== SPELLS.WINDBURST.id) {
+      return;
+    }
+    if (event.classResources) {
+      event.classResources
+        .filter(resource => resource.type === RESOURCE_TYPES.FOCUS)
+        .forEach(({ cost }) => {
+          const focusCost = cost || 0;
+          this.focusUpdates.unshift({
```

Comment:
I'm trying to think about whether the array is really necessary. You alter it by `unshift`ing each affected spell's Focus cost, so on `[0]` you always have the most recent one, which you also use on lines 45 and 47, but you never really use the array any other way, like iterating through it, doing come math on it etc. Couldn't this be replaced by something like `this.lastFocusCost`?

Reply:
It probably could, I can test a few things post raid - this was just a working solution 😊

Reply:
Sure thing, it's a good thing to have something that works :)